### PR TITLE
hdmi_cec: Test Profile: Adding MACRO ENABLE_ENHANCED_ERROR_CODE to enalbe/disable the enhanced error code tests

### PR DESCRIPTION
**Pull request by akhilbhas**
Adding Macro ENABLE_ENHANCED_ERROR_CODE to enable/disable enhanced error code test and test step in l1

### CODE REVIEW

**Review by srinivasgtl — COMMENTED** [review]
Also the following changes needs to be handled.
Line 336 - HDMI_CEC_IO_LOGICALADDRESS_UNAVAILABLE  should be commented.. and entire function should be commented.. and later moved to l3.
HDMI_CEC_IO_NOT_OPENED is a enhanced API as per Interface file.
HDMI_CEC_IO_INVALID_HANDLE is also part of enhanced error code. 

Please check for the above two error codes which are not part of older error codes and rectify.  There are too many places where these error codes needs to be disabled.  

for the older error code check the tags 2.0 version (https://github.com/rdkcentral/rdk-halif-hdmi_cec/blob/0.2.0/include/hdmi_cec_driver.h#L293)

File: src/test_l1_hdmi_cec_driver.c [hunk 1]
```diff
@@ -467,16 +479,21 @@ void test_hdmicec_hal_l1_getPhysicalAddress_negative( void )
     gTestID = 6;
 
     UT_LOG("\n In %s [%02d%03d]\n", __FUNCTION__, gTestGroup, gTestID);
+#ifdef ENABLE_ENHANCED_ERROR_CODE
```
[the remark this answers is not in the document]
Comment (akhilbhas) [reply src/test_l1_hdmi_cec_driver.c]: Added else part.

File: src/test_l1_hdmi_cec_driver.c [hunk 2]
```diff
@@ -494,9 +511,11 @@ void test_hdmicec_hal_l1_getPhysicalAddress_negative( void )
     result = HdmiCecClose (handle);
     if (HDMI_CEC_IO_SUCCESS != result) { UT_FAIL ("close failed"); }
 
+#ifdef ENABLE_ENHANCED_ERROR_CODE
     //Calling API after close,
     result = HdmiCecGetPhysicalAddress(handle, &physicalAddress);
     if (HDMI_CEC_IO_SUCCESS  != result) { UT_FAIL ("HdmiCecGetPhysicalAddress failed"); }
```
[the remark this answers is not in the document]
Comment (akhilbhas) [reply src/test_l1_hdmi_cec_driver.c]: Added

File: src/test_l1_hdmi_cec_driver.c [hunk 3]
```diff
@@ -843,13 +882,19 @@ void test_hdmicec_hal_l1_removeLogicalAddress_sinkDevice_negative( void )
     if (HDMI_CEC_IO_SUCCESS != result) { HDMICEC_ASSERT_AUTO_TERM(); UT_FAIL ("HdmiCecRemoveLogicalAddress failed"); }
 
     result = HdmiCecRemoveLogicalAddress( handle, logicalAddress );
+#ifdef ENABLE_ENHANCED_ERROR_CODE
```
[the remark this answers is not in the document]
Comment (akhilbhas) [reply src/test_l1_hdmi_cec_driver.c:930]: Addressed.

File: src/test_l1_hdmi_cec_driver.c [hunk 4]
```diff
@@ -814,9 +852,10 @@ void test_hdmicec_hal_l1_removeLogicalAddress_sinkDevice_negative( void )
     gTestID = 12;
 
     UT_LOG("\n In %s [%02d%03d]\n", __FUNCTION__, gTestGroup, gTestID);
+#ifdef ENABLE_ENHANCED_ERROR_CODE
     result = HdmiCecRemoveLogicalAddress(handle, logicalAddress );
     if (HDMI_CEC_IO_NOT_OPENED != result) { UT_FAIL ("HdmiCecRemoveLogicalAddress failed"); }
-
+#endif
```
[the remark this answers is not in the document]
Comment (akhilbhas) [reply src/test_l1_hdmi_cec_driver.c:896]: Addressed

**Comment by akhilbhas** [comment]
Updated and verified all the following Enhanced error codes are handled
* HDMI_CEC_IO_LOGICALADDRESS_UNAVAILABLE - the Line where the tests are added to the suite is commented out.
* HDMI_CEC_IO_NOT_OPENED - Added macro and else with HDMI_CEC_IO_INVALID_ARGUMENT as handle is Invalid to the fucntion
* HDMI_CEC_IO_INVALID_HANDLE - Added Macro and else with HDMI_CEC_IO_INVALID_ARGUMENT 

HDMI_CEC_IO_INVALID_HANDLE:  For _"test_hdmicec_hal_l1_open_negative" this path is not handled when passing handle as NULL to HdmiCecOpen, It will result in crash, hence commented the call itself to avoid the crash.

File: src/test_l1_hdmi_cec_driver.c [hunk 5]
```diff
@@ -245,14 +245,19 @@ void test_hdmicec_hal_l1_open_negative( void )
 
     UT_LOG("\n In %s [%02d%03d]\n", __FUNCTION__, gTestGroup, gTestID);
     //Check Null even before calling the positive case
+#ifdef ENABLE_ENHANCED_ERROR_CODE
     result = HdmiCecOpen( NULL );
     if (HDMI_CEC_IO_INVALID_HANDLE  != result) { UT_FAIL("open failed"); }
```
Comment (Ulrond) [inline src/test_l1_hdmi_cec_driver.c:250]: if (HDMI_CEC_IO_INVALID_HANDLE  != result) { UT_FAIL("open failed"); }

should be

```c
UT_ASSERT_EQUAL( (HDMI_CEC_IO_INVALID_HANDLE, result );
or
UT_ASSERT( (HDMI_CEC_IO_INVALID_HANDLE == result );
```

Comment (Ulrond) [reply src/test_l1_hdmi_cec_driver.c:250]:
```c
    result = HdmiCecOpen( &handle );
#ifdef ENABLE_ENHANCED_ERROR_CODE
    if (HDMI_CEC_IO_ALREADY_OPEN != result) { UT_FAIL("open failed"); }
#else
    if (HDMI_CEC_IO_SUCCESS != result) { UT_FAIL("open failed"); }
#endif
```

becomes

```
    result = HdmiCecOpen( &handle );
    if ( feature_enchanged_error_code )
    {
       UT_ASSERT_EQUALS( HDMI_CEC_IO_ALREADY_OPEN, result );
   }
   else
   {
       UT_ASSERT_EQUALS( HDMI_CEC_IO_SUCCESS, result );
   }
```

becomes

```c
 void checkEnchancedError( result, enhanced, old )
{
      if ( feature_enchanged_error_code )
    {
       UT_ASSERT_EQUALS( enhanced, result );
   }
   else
   {
       UT_ASSERT_EQUALS( old, result );
   }
}
```

```c
    result = HdmiCecOpen( &handle );
   checkEnchancedError( result, HDMI_CEC_IO_ALREADY_OPEN, HDMI_CEC_IO_SUCCESS );
```

File: src/test_l1_hdmi_cec_driver.c [hunk 6]
```diff
@@ -245,14 +245,19 @@ void test_hdmicec_hal_l1_open_negative( void )
 
     UT_LOG("\n In %s [%02d%03d]\n", __FUNCTION__, gTestGroup, gTestID);
     //Check Null even before calling the positive case
+#ifdef ENABLE_ENHANCED_ERROR_CODE
```
Comment (Ulrond) [inline src/test_l1_hdmi_cec_driver.c:248]: please update to this requirement -> https://github.com/rdkcentral/ut-core/discussions/55

**Review by Ulrond — CHANGES_REQUESTED** [review]
https://github.com/rdkcentral/ut-core/discussions/55 -> please update your requirement to support KVP module.

File: src/test_l1_hdmi_cec_driver.c (hunk 6 above)
Comment (akhilbhas) [reply src/test_l1_hdmi_cec_driver.c]: Updated the code as per the suggetion.

File: ut/profile/hdmiCECExtendedEnumsNotSupported.yaml [hunk 7]
```diff
@@ -0,0 +1,3 @@
+hdmicec:
```
Comment (Ulrond) [inline ut/profile/hdmiCECExtendedEnumsNotSupported.yaml:1]: this is in the wrong place.

Look at your directory structure when you clone this repo.

header(rdk-halif-hdmi_cec)->ut(rdk-halif-test-hdmi_cec)->ut_core

So this directory isn't `ut/profile` from this git repo point of view it's `profile`

File: src/test_l1_hdmi_cec_driver.c [hunk 8]
```diff
@@ -68,6 +68,7 @@
 
 #include <ut.h>
 #include "ut_log.h"
+#include "ut_kvp_profile.h"
```
Comment (Ulrond) [inline src/test_l1_hdmi_cec_driver.c:71]: techically include "file.h" means search the local local directory then Path, where include <file.h>, means look in the path.

so #include <file.h> is more correct.

File: src/test_l1_hdmi_cec_driver.c [hunk 9]
```diff
@@ -596,31 +606,34 @@ void test_hdmicec_hal_l1_addLogicalAddress_sinkDevice_negative( void )
     UT_LOG("\n In %s [%02d%03d]\n", __FUNCTION__, gTestGroup, gTestID);
 
     result = HdmiCecAddLogicalAddress( handle, logicalAddress );
+#ifdef EHNCAGED_CHECKS_ENABLED
```
Comment (Ulrond) [inline src/test_l1_hdmi_cec_driver.c:609]: why is this one different that all the others?

**Review by Ulrond — CHANGES_REQUESTED** [review]
comments to review.

File: src/test_l1_hdmi_cec_driver.c (hunk 8 above)
Comment (akhilbhas) [reply src/test_l1_hdmi_cec_driver.c]: Updated the lines
#include "ut_log.h"
#include "ut_kvp_profile.h"

File: src/test_l1_hdmi_cec_driver.c (hunk 9 above)
Comment (akhilbhas) [reply src/test_l1_hdmi_cec_driver.c]: This line was missed when updating.
Updated this line as well.

File: src/test_l1_hdmi_cec_driver.c [hunk 10]
```diff
@@ -246,13 +259,13 @@ void test_hdmicec_hal_l1_open_negative( void )
     UT_LOG("\n In %s [%02d%03d]\n", __FUNCTION__, gTestGroup, gTestID);
     //Check Null even before calling the positive case
     result = HdmiCecOpen( NULL );
-    if (HDMI_CEC_IO_INVALID_HANDLE  != result) { UT_FAIL("open failed"); }
+    CHECK_FOR_EXTENDED_ERROR_CODE(result,HDMI_CEC_IO_INVALID_HANDLE,HDMI_CEC_IO_INVALID_ARGUMENT)
```
Comment (Ulrond) [inline src/test_l1_hdmi_cec_driver.c:262]: does this work? we should add ; on the end

**Review by Ulrond — DISMISSED** [review]
looks ok except for comment about ; ( which I think you should fix either way ).

Can you post the results of a test run just to check that it's working ok?

File: src/test_l1_hdmi_cec_driver.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ulrond) [reply src/test_l1_hdmi_cec_driver.c]: resolve conversations if you've fixed the problem.

File: src/test_l1_hdmi_cec_driver.c [hunk 11]
```diff
@@ -341,7 +354,7 @@ void test_hdmicec_hal_l1_open_logical_address_unavailable_source ( void )
 
      //Here handle = 0 since open failed and close should fail.
     result = HdmiCecClose( handle );
-    if (HDMI_CEC_IO_INVALID_HANDLE  != result) { UT_FAIL_FATAL("close failed"); }
+    CHECK_FOR_EXTENDED_ERROR_CODE(result,HDMI_CEC_IO_INVALID_HANDLE,HDMI_CEC_IO_INVALID_ARGUMENT)
```
Comment (srinivasgtl) [inline src/test_l1_hdmi_cec_driver.c:357]: Let's remove this entire test case as this requires external devices to be connected.  We shall move this test to L3

File: src/test_l1_hdmi_cec_driver.c [hunk 12]
```diff
@@ -475,15 +487,14 @@ void test_hdmicec_hal_l1_getPhysicalAddress_negative( void )
...
-
+    CHECK_FOR_EXTENDED_ERROR_CODE(result,HDMI_CEC_IO_NOT_OPENED,HDMI_CEC_IO_GENERAL_ERROR)
 
     result = HdmiCecOpen (&handle);
     //if init is failed no need to proceed further
     UT_ASSERT_EQUAL_FATAL( result, HDMI_CEC_IO_SUCCESS );
 
     result = HdmiCecGetPhysicalAddress(0, &physicalAddress);
-    if (HDMI_CEC_IO_INVALID_HANDLE  != result) { UT_FAIL("HdmiCecGetPhysicalAddress failed"); }
+    CHECK_FOR_EXTENDED_ERROR_CODE(result,HDMI_CEC_IO_INVALID_HANDLE,HDMI_CEC_IO_INVALID_HANDLE)
```
Comment (srinivasgtl) [inline src/test_l1_hdmi_cec_driver.c:497]: At line 506, as per the new interface file, let's compare with F.F.F.F and not with 4.4.4.4

File: src/test_l1_hdmi_cec_driver.c [hunk 13]
```diff
@@ -596,31 +606,30 @@ void test_hdmicec_hal_l1_addLogicalAddress_sinkDevice_negative( void )
...
-    if (HDMI_CEC_IO_INVALID_HANDLE  != result) { UT_FAIL("HdmiCecAddLogicalAddress failed"); }
+    CHECK_FOR_EXTENDED_ERROR_CODE(result,HDMI_CEC_IO_INVALID_HANDLE,HDMI_CEC_IO_INVALID_ARGUMENT)
 
     result = HdmiCecAddLogicalAddress( handle, -1 );
-    if (HDMI_CEC_IO_INVALID_ARGUMENT  != result) { UT_FAIL("HdmiCecAddLogicalAddress failed"); }
+    CHECK_FOR_EXTENDED_ERROR_CODE(result,HDMI_CEC_IO_INVALID_ARGUMENT,HDMI_CEC_IO_GENERAL_ERROR)
 
     result = HdmiCecAddLogicalAddress( handle, 0x3 );
-    if (HDMI_CEC_IO_INVALID_ARGUMENT  != result) { UT_FAIL("HdmiCecAddLogicalAddress failed"); }
+    CHECK_FOR_EXTENDED_ERROR_CODE(result,HDMI_CEC_IO_INVALID_ARGUMENT,HDMI_CEC_IO_GENERAL_ERROR)
```
Comment (srinivasgtl) [inline src/test_l1_hdmi_cec_driver.c:623]: As per new hdmi cec interface file, we can add any value between 0x00 to 0x0f, so this test and the below one should get passed.   delete the test cases here

File: src/test_l1_hdmi_cec_driver.c [hunk 14]
```diff
@@ -712,14 +721,14 @@ void test_hdmicec_hal_l1_addLogicalAddress_sourceDevice( void )
 
     UT_LOG("\n In %s [%02d%03d]\n", __FUNCTION__, gTestGroup, gTestID);
     result = HdmiCecAddLogicalAddress( handle, logicalAddress );
-    if (HDMI_CEC_IO_NOT_OPENED  != result) { UT_FAIL("HdmiCecAddLogicalAddress failed"); }
-
+    CHECK_FOR_EXTENDED_ERROR_CODE(result,HDMI_CEC_IO_NOT_OPENED,HDMI_CEC_IO_INVALID_ARGUMENT)
```
Comment (srinivasgtl) [inline src/test_l1_hdmi_cec_driver.c:724]: As per new interface, the add logical address for a source device should return UN SUPPORTED. review entire test case.

File: src/test_l1_hdmi_cec_driver.c [hunk 15]
```diff
@@ -848,7 +857,7 @@ void test_hdmicec_hal_l1_removeLogicalAddress_sinkDevice_negative( void )
     if (HDMI_CEC_IO_SUCCESS != result) { UT_FAIL("HdmiCecRemoveLogicalAddress failed"); }
 
     result = HdmiCecRemoveLogicalAddress( handle, logicalAddress );
-    if (HDMI_CEC_IO_ALREADY_REMOVED != result) { UT_FAIL("HdmiCecRemoveLogicalAddress failed"); }
+    CHECK_FOR_EXTENDED_ERROR_CODE(result,HDMI_CEC_IO_ALREADY_REMOVED,HDMI_CEC_IO_SUCCESS)
```
Comment (srinivasgtl) [inline src/test_l1_hdmi_cec_driver.c:860]: Remove Steps 856 and 857.  Removing without adding should not be a success.  These steps are not required any more.

File: src/test_l1_hdmi_cec_driver.c [hunk 16]
```diff
@@ -954,7 +963,7 @@ void test_hdmicec_hal_l1_removeLogicalAddress_sourceDevice( void )
     UT_LOG("\n In %s [%02d%03d]\n", __FUNCTION__, gTestGroup, gTestID);
 
     result = HdmiCecRemoveLogicalAddress(handle, logicalAddress );
-    if (HDMI_CEC_IO_NOT_OPENED != result) { UT_FAIL("HdmiCecRemoveLogicalAddress failed"); }
+    CHECK_FOR_EXTENDED_ERROR_CODE(result,HDMI_CEC_IO_NOT_OPENED,HDMI_CEC_IO_INVALID_ARGUMENT)
```
Comment (srinivasgtl) [inline src/test_l1_hdmi_cec_driver.c:966]: As per new interface file, these functions should return NOT SUPPORTED from the source device. check for not supported.

File: src/test_l1_hdmi_cec_driver.c [hunk 17]
```diff
@@ -1801,7 +1809,7 @@ void test_hdmicec_hal_l1_hdmiCecTxAsync_sinkDevice_negative( void )
     UT_LOG("\n In %s [%02d%03d]\n", __FUNCTION__, gTestGroup, gTestID);
     //Calling API before open, should give invalid argument
     result = HdmiCecTxAsync(handle, buf, len); //Code crash here
-    if (HDMI_CEC_IO_NOT_OPENED != result) { UT_FAIL("HdmiCecTxAsync failed"); }
+    CHECK_FOR_EXTENDED_ERROR_CODE(result,HDMI_CEC_IO_NOT_OPENED,HDMI_CEC_IO_INVALID_ARGUMENT)
```
Comment (srinivasgtl) [inline src/test_l1_hdmi_cec_driver.c:1812]: Async is now completely removed. So this entire function should be removed for both sink and source devices.

File: src/test_l1_hdmi_cec_driver.c [hunk 18]
```diff
@@ -2007,7 +2015,7 @@ void test_hdmicec_hal_l1_hdmiCecTxAsync_sourceDevice_negative( void )
 
     //Calling API after close, should give invalid argument
     result = HdmiCecTxAsync(handle, buf, len);
-    if (HDMI_CEC_IO_NOT_OPENED != result) { UT_FAIL("HdmiCecTxAsync failed"); }
+    CHECK_FOR_EXTENDED_ERROR_CODE(result,HDMI_CEC_IO_NOT_OPENED,HDMI_CEC_IO_INVALID_ARGUMENT)
```
Comment (srinivasgtl) [inline src/test_l1_hdmi_cec_driver.c:2018]: Remove all AsyncTx tests and also the port disconnected tests which are part of L3 again.

**Review by srinivasgtl — COMMENTED** [review]
Please update the following test review comments.

File: src/test_l1_hdmi_cec_driver.c (hunk 10 above)
Comment (akhilbhas) [reply src/test_l1_hdmi_cec_driver.c]: Added ; to all the Macro calls

File: src/test_l1_hdmi_cec_driver.c (hunk 18 above)
Comment (srinivasgtl) [reply src/test_l1_hdmi_cec_driver.c]: need to consider as another PR.

File: src/test_l1_hdmi_cec_driver.c (hunk 17 above)
Comment (srinivasgtl) [reply src/test_l1_hdmi_cec_driver.c]: It will be considered as a separate PR.

File: src/test_l1_hdmi_cec_driver.c (hunk 15 above)
Comment (srinivasgtl) [reply src/test_l1_hdmi_cec_driver.c]: will consider in a different PR

File: src/test_l1_hdmi_cec_driver.c (hunk 14 above)
Comment (srinivasgtl) [reply src/test_l1_hdmi_cec_driver.c]: Will consider as a different PR.

File: src/test_l1_hdmi_cec_driver.c (hunk 13 above)
Comment (srinivasgtl) [reply src/test_l1_hdmi_cec_driver.c]: will consider as a different PR.

File: src/test_l1_hdmi_cec_driver.c (hunk 11 above)
Comment (srinivasgtl) [reply src/test_l1_hdmi_cec_driver.c]: Will consider this in a different PR.

File: src/test_l1_hdmi_cec_driver.c (hunk 12 above)
Comment (srinivasgtl) [reply src/test_l1_hdmi_cec_driver.c]: It will be considered in a different PR.

**Review by srinivasgtl — APPROVED** [review]
approved.

**Pull request merged**